python/lsst/daf/butler/tests/butler_query.py

```diff
@@ -0,0 +1,1228 @@
+            Test(
+                "tract,visit", dataId={"instrument": "Cam1", "skymap": "SkyMap2"}, exception=DataIdValueError
+            ),
+            Test("tract,visit", where="instrument='Cam1' AND skymap='SkyMap1'", count=6),
+            Test("tract,visit", where="instrument='Cam1' AND skymap='SkyMap5'", exception=DataIdValueError),
+            Test(
+                "tract,visit",
+                where="instrument=cam AND skymap=map",
+                bind={"cam": "Cam1", "map": "SkyMap1"},
+                count=6,
+            ),
+            Test(
+                "tract,visit",
+                where="instrument=cam AND skymap=map",
+                bind={"cam": "Cam", "map": "SkyMap"},
+                exception=DataIdValueError,
+            ),
+        )
+
+        for test in test_data:
+            dimensions = test.dimensions.split(",")
+            if test.exception:
+                with self.assertRaises(test.exception):
+                    do_query(dimensions, test.dataId, test.where, bind=test.bind, **test.kwargs).count()
+            else:
+                query = do_query(dimensions, test.dataId, test.where, bind=test.bind, **test.kwargs)
+                self.assertEqual(query.count(discard=True), test.count)
+
+            # and materialize
+            if test.exception:
+                with self.assertRaises(test.exception):
+                    query = do_query(dimensions, test.dataId, test.where, bind=test.bind, **test.kwargs)
+                    with query.materialize() as materialized:
+                        materialized.count(discard=True)
+            else:
+                query = do_query(dimensions, test.dataId, test.where, bind=test.bind, **test.kwargs)
+                with query.materialize() as materialized:
+                    self.assertEqual(materialized.count(discard=True), test.count)
+
+    def testQueryDimensionRecordsExceptions(self) -> None:
+        """Test exceptions raised by query.dimension_records()."""
+        butler = self.make_butler("base.yaml", "datasets.yaml", "spatial.yaml")
+
+        result = butler.query.dimension_records("detector")
+        self.assertEqual(result.count(), 4)
+        result = butler.query.dimension_records("detector", instrument="Cam1")
+        self.assertEqual(result.count(), 4)
+        result = butler.query.dimension_records("detector", data_id={"instrument": "Cam1"})
+        self.assertEqual(result.count(), 4)
+        result = butler.query.dimension_records("detector", where="instrument='Cam1'")
+        self.assertEqual(result.count(), 4)
+        result = butler.query.dimension_records("detector", where="instrument=instr", bind={"instr": "Cam1"})
+        self.assertEqual(result.count(), 4)
+
+        with self.assertRaisesRegex(DataIdValueError, "dimension instrument"):
+            result = butler.query.dimension_records("detector", instrument="NotCam1")
+            result.count()
```

Comment:
Another case where we might want to move the `with self.assertRaisesRegex` a line lower to make sure the exception is happening where we expect.  Either that or the test is bad.

Lots more of these below.  It's possible the original tests got this wrong, and the behavior we're getting is what we want even if it's not what we're testing for.

Reply:
Call to `count()` is not really needed to trigger an exception. Maybe things changed since from when the test was originally written. I'll drop all `count` calls.